appserver/payara-appserver-modules/payara-micro-service/src/main/java/fish/payara/appserver/context/JavaEEContextUtilImpl.java

```diff
@@ -0,0 +1,170 @@
+        InvocationManager invMgr = serverContext.getInvocationManager();
+        boolean invocationCreated = false;
+        if(invMgr.getCurrentInvocation() == null) {
+            invMgr.preInvoke(new ComponentInvocation(capturedInvocation.getComponentId(), ComponentInvocation.ComponentInvocationType.SERVLET_INVOCATION,
+                    capturedInvocation.getContainer(), capturedInvocation.getAppName(), capturedInvocation.getModuleName()));
+            invocationCreated = true;
+        }
+        JndiNameEnvironment componentEnv = compEnvMgr.getCurrentJndiNameEnvironment();
+        if(invocationCreated && componentEnv instanceof BundleDescriptor) {
+            BundleDescriptor bd = (BundleDescriptor)componentEnv;
+            Utility.setContextClassLoader(bd.getClassLoader());
+        }
+        return new ContextImpl.Context(oldClassLoader, invocationCreated? invMgr.getCurrentInvocation() : null);
+    }
+
+    @Override
+    public void postInvoke(Context _ctx) {
+        ContextImpl.Context ctx = (ContextImpl.Context)_ctx;
+        if (ctx.invocation != null) {
+            getServerContext().getInvocationManager().postInvoke(ctx.invocation);
+            Utility.setContextClassLoader(ctx.classLoader);
+        }
+    }
+
+    /**
+     * pushes invocation context onto the stack
+     * Also creates Request scope
+     *
+     * @return new context that was created
+     */
+    @Override
+    public RequestContext preInvokeRequestContext() {
+        Context rootCtx = preInvoke();
+        BoundRequestContext brc = CDI.current().select(BoundRequestContext.class).get();
+        ContextImpl.RequestContext context = new ContextImpl.RequestContext(rootCtx, brc.isActive()? null : brc, new HashMap<String, Object>());
+        if(context.ctx != null) {
+            context.ctx.associate(context.storage);
+            context.ctx.activate();
+        }
+        return context;
+    }
+
+    /**
+     * context to pop from the stack
+     *
+     * @param context to be popped
+     */
+    @Override
+    public void postInvokeRequestContext(RequestContext context) {
+        ContextImpl.RequestContext ctx = (ContextImpl.RequestContext)context;
+        if (ctx.ctx!= null) {
+            ctx.ctx.deactivate();
+            ctx.ctx.dissociate(ctx.storage);
+        }
+        postInvoke(ctx.rootCtx);
+    }
+
+    /**
+     * @return application name or null if there is no invocation context
+     */
+    @Override
+    public String getApplicationName() {
+        ComponentInvocation ci = serverContext.getInvocationManager().getCurrentInvocation();
+        return ci != null? ci.getAppName() : null;
+    }
+
+    /**
+     * set context class loader by appName
+     *
+     * @param appName
+     */
+    @Override
+    public void setApplicationContext(String appName) {
+        if(appName == null) {
+            return;
+        }
+        ApplicationInfo appInfo = appRegistry.get(appName);
+        if(appInfo != null) {
```

Comment:
Also, if the app is undeployed, this may not result in a failure because it may not use the app-specific classes altogether.  If this fails, it will result in ClassNotFound exception, which will clue in the folks about the issue.

In other words, it's is all good, and not even a warning is necessary